mods/pageviews.js

```diff
@@ -0,0 +1,277 @@
+        attributes: {
+            project: 'string',
+            access: 'string',
+            agent: 'string',
+            granularity: 'string',
+            // the hourly timestamp will be stored as YYYYMMDDHH
+            timestamp: 'string',
+            views: 'int'
+        },
+        index: [
+            { attribute: 'project', type: 'hash' },
+            { attribute: 'access', type: 'hash' },
+            { attribute: 'agent', type: 'hash' },
+            { attribute: 'granularity', type: 'hash' },
+            { attribute: 'timestamp', type: 'range', order: 'asc' },
+        ]
+    },
+    tops: {
+        table: tables.tops,
+        version: 1,
+        attributes: {
+            project: 'string',
+            access: 'string',
+            year: 'string',
+            month: 'string',
+            day: 'string',
+            // format for this is a json array: [{rank: 1, article: <<title>>, views: 123}, ...]
+            articles: 'string'
+        },
+        index: [
+            { attribute: 'project', type: 'hash' },
+            { attribute: 'access', type: 'hash' },
+            { attribute: 'year', type: 'hash' },
+            { attribute: 'month', type: 'hash' },
+            { attribute: 'day', type: 'hash' },
+        ]
+    }
+};
+
+/**
+ * general handler functions */
+var queryResponser = function(res) {
+    // always return at least an empty array so that queries for non-existing data don't error
+    res = res || {};
+    res.body = res.body || { items: [] };
+    res.headers = res.headers || {};
+    // NOTE: We decided to let "data not found" be reported as a 404.  We have a work-around if
+    // consumers complain that they prefer a 204 instead:
+    //   We could catch the 404 and run the same query without the date parameters.  If we find
+    //   results (use limit 1 for efficiency), we could then return a 204 because we'd know the
+    //   dates were the part of the query that wasn't found.
+    return res;
+};
+
+/**
+ * Parameter validators
+ */
+var throwIfNeeded = function(errors) {
+    if (errors && errors.length) {
+        throw new rbUtil.HTTPError({
+            status: 400,
+            body: {
+                type: 'invalid_request',
+                detail: errors,
+            }
+        });
+    }
+};
+
+/**
+ * Cleans the project parameter so it can be passed in as either en.wikipedia.org or en.wikipedia
+ */
+var stripOrgFromProject = function(rp) {
+    rp.project = rp.project.replace(/^(.*).org$/, '$1');
+};
+
+var validateStartAndEnd = function(rp) {
+    var errors = [];
+
+    stripOrgFromProject(rp);
+
+    if (!/^[0-9]{10}$/.test(rp.start)) {
+        errors.push('start timestamp is invalid');
+    }
+    if (!/^[0-9]{10}$/.test(rp.end)) {
+        errors.push('end timestamp is invalid');
+    }
+
+    throwIfNeeded(errors);
+};
+
+var validateYearMonthDay = function(rp) {
+    var errors = [];
+
+    stripOrgFromProject(rp);
+
+    if (rp.year === 'all-years') {
```

Comment:
I'll document this better in the parameters, to explain the behavior.  But we didn't want to go down the weird path of returning the month of July for "all years".  So is that ok if the behavior is explained but stays the same?